Widget

sxzhang25 opened the pull request:
Video Player widget for motion photos.

sxzhang25 committed:
add default files

sxzhang25 committed:
Add default files.

sxzhang25 committed:
add default files

sxzhang25 committed:
Add descriptive log messages.

sxzhang25 committed:
Remove build and .gradle folders.

sxzhang25 committed:
Move widget setup into separate method.

sxzhang25 committed:
Merge branch 'motion-photo-reader' of https://github.com/googleinterns/motion-photos-filereader into widget-copy

Pull latest changes to motion photo reader files.

sxzhang25 committed:
Refactor to start playback thread after surface is created.

sxzhang25 committed:
Clean up code.

sxzhang25 committed:
Clean up code and ignore build files.

sxzhang25 committed:
Add method to get current video timestamp.

sxzhang25 committed:
Add MotionPhotoWidget unit tests.

sxzhang25 committed:
Switch from SurfaceView to TextureView in order to preserve state after screen orientation change.

sxzhang25 committed:
Add saved state for video player to use.

sxzhang25 committed:
Add saved state for video player to use.

sxzhang25 committed:
Add saved state for video player to use.

sxzhang25 committed:
Add saved state for video player to use.

sxzhang25 committed:
Merge remote-tracking branch 'origin/widget-copy' into widget-copy

# Conflicts:
#	MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/MotionPhotoWidget.java

sxzhang25 committed:
Enable player to continue in saved state on phone rotation.

sxzhang25 committed:
Enable player to continue in saved state on phone rotation.

sxzhang25 committed:
Merge remote-tracking branch 'origin/widget-copy' into widget-copy

# Conflicts:
#	MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/MotionPhotoWidget.java

sxzhang25 committed:
Add argument annotations.

sxzhang25 committed:
Enable widget to continue playing in previous state on phone rotation.

sxzhang25 committed:
fix merge conflicts from master

sxzhang25 committed:
Fix isEmpty().

Signed-off-by: sxzhang25 <zsharonx@gmail.com>

sxzhang25 committed:
Add Espresso unit tests.

Signed-off-by: sxzhang25 <zsharonx@gmail.com>

sxzhang25 committed:
Refactor SurfaceTextureListener into a custom WidgetSurfaceTextureListener class for readability.

rjuang commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/MotionPhotoWidget.java:
```diff
@@ -0,0 +1,279 @@
+        }
+
+        public static final Parcelable.Creator<SavedState> CREATOR
+                = new Parcelable.Creator<SavedState>() {
```
nit, move the `=` operator to line above with CREATOR... assignments seem to be the only exception to the rule for having the operator start the new line.

rjuang commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/MotionPhotoWidget.java:
```diff
@@ -0,0 +1,279 @@
+public class MotionPhotoWidget extends TextureView {
+
+    private static final String TAG = "MotionPhotoWidget";
+    private static final Object lock = new Object();
```
unused variable. delete line?

rjuang commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/MotionPhotoWidget.java:
```diff
@@ -0,0 +1,279 @@
+        setup();
+    }
+
+    private void setup() {
```
nit: rename this to initialize() 

Feels too close to  setUp that's used for testbenches.

rjuang commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/MotionPhotoWidget.java:
```diff
@@ -0,0 +1,316 @@
+        }
+
+        @Override
+        public void onSurfaceTextureUpdated(SurfaceTexture surface) {
+
+        }
```
condense this to a single line, i.e.

```
@Override
public void onSurfaceTextureUpdated(SurfaceTexture surface) {}
```

sxzhang25 committed:
Account for video orientation when rendering to view.

rjuang commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/MotionPhotoWidget.java:
```diff
@@ -0,0 +1,316 @@
+    private void setup() {
+        // Set up the executor and play/pause process to facilitate stopping and starting the video
+        playProcess = new PlayProcess();
+        executor = Executors.newSingleThreadExecutor(new ThreadFactory() {
+            @Override
+            public Thread newThread(Runnable r) {
+                return new Thread(r) {
+                    @Override
+                    public void interrupt() {
+                        super.interrupt();
+                        playProcess.cancel();
+                    }
+                };
+            }
+        });
```
This is an interesting idea but might be overkill for this. You should just directly call `playProcess.cancel()` in the associated close/shutdown (which is also missing).

This executor will leak everytime it is created. Also, `executor.interrupt` nor `executor.shutdown`  is never called, so this thread will never be interrupted. Consider either overriding the view's `onDetachFromWindow` and `onAttachToWindow` to setup the begin and end lifecylce.

A cleaner approach might be to migrate to `Fragment` since those do properly implement a create/destroy lifecycle

sxzhang25 committed:
Add in initial playback speed code and simplify play/pause control in widget.

Moved executor creation to view attachmment and add executor shutdown to view detachment to prevent leaks.

sxzhang25 committed:
Merge branch 'master' of https://github.com/googleinterns/motion-photos-filereader into widget-copy

Update MotionPhotoReader to include added support for motion photo v2
formats.

sxzhang25 committed:
Remove background color attribute from widget, and set it in XML layout file  instead.

sxzhang25 committed:
Remove background color attribute from widget, and set it in XML layout file  instead.

sxzhang25 committed:
Merge branch 'widget-copy' of https://github.com/googleinterns/motion-photos-filereader into widget-copy

sxzhang25 committed:
Add fill option to widget.

When set to true, the fill option scales the video to fill the entire surface texture. Otherwise, the video is scaled to fit in the surface texture. By default, the fill option is set to false.

sxzhang25 committed:
Override visibility changes to pause reader from running when widget is backgrounded.

sxzhang25 committed:
Fix playback speed.

Add fields into BufferProcessor which store render timestamps, allowing the processer to compute frame deltas and post render timestamps to the output buffers that reflect the fram rate dictated by video timestamps.

sxzhang25 commented:
Current playback speed changes are not working yet. I'm putting up the code (the playback section is not polished yet) so that there is a copy of what I currently have, for reference.

sxzhang25 committed:
Use SystemClock instead of System and update activity layout.

sxzhang25 committed:
Switch back to SurfaceView.

Switch from TextureView to SurfaceView in order to allow for rendering timestamps to work.

Signed-off-by: sxzhang25 <zsharonx@gmail.com>

sxzhang25 committed:
Fix state continuations of surface.

- Close reader when view is detached.
- Allow video to revive after app is backgrounded and reopened.

rjuang commented on MotionPhotoReader/app/src/androidTest/java/com/google/android/libraries/motionphotoreader/MotionPhotoWidgetTest.java:
```diff
@@ -0,0 +1,125 @@
+    @Test
+    public void widgetUI_surface_isPrepared() throws InterruptedException {
+        ViewInteraction motionPhotoWidget = onView(withId(R.id.motion_photo_widget));
+        Thread.sleep(1000);
+    }
```
I'm wondering if there's a way to validate that these happened ? It looks like this will guard against crashes, but won't guarantee if the interaction occurred ?

rjuang commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/BufferProcessor.java:
```diff
@@ -136,9 +142,33 @@ public void process(Bundle messageData) {
+                long renderTimestampNs;
+                if (prevRenderTimestampNs == 0) {
+                    renderTimestampNs =
+                            System.nanoTime() + frameDeltaUs * US_TO_NS;
```
can we assign System.nanoTime() to a variable so that the value isn't changing between each use and comparison? e.g.,

```
long currentTimestampNs = System.nanoTime();

...
renderTimestampNs = currentTimestampNs + frameDeltaUs  * US_TO_NS;
```

rjuang commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/BufferProcessor.java:
```diff
@@ -136,9 +142,33 @@ public void process(Bundle messageData) {
+                } else {
+                    renderTimestampNs = prevRenderTimestampNs + frameDeltaUs * US_TO_NS;
+                }
+                if (renderTimestampNs < System.nanoTime()) {
```
substitute System.nanoTime with currentTimestampNs (mentioned in earlier comment)

rjuang commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/BufferProcessor.java:
```diff
@@ -136,9 +142,33 @@ public void process(Bundle messageData) {
+                if (renderTimestampNs < System.nanoTime()) {
+                    Log.d(TAG, "render timestamp adjusted");
+                    renderTimestampNs = System.nanoTime() + frameDeltaUs * US_TO_NS;
+                }
```
replace System.nanoTime with currentTimestampNs (mentioned in earlier comment)

rjuang commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/MainActivity.java:
```diff
@@ -1,6 +1,61 @@
+import android.view.View;
+
+import androidx.annotation.RequiresApi;
+
```
nit: remove blank line

rjuang commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/MainActivity.java:
```diff
@@ -1,6 +1,61 @@
+import androidx.annotation.RequiresApi;
+
+import com.adobe.internal.xmp.XMPException;
+
```
nit: remove blank line

rjuang commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/MainActivity.java:
```diff
@@ -1,6 +1,61 @@
+        motionPhotoWidget.setOnClickListener(new MyOnWidgetClickListener(motionPhotoWidget));
+    }
+
+    private class MyOnWidgetClickListener implements View.OnClickListener {
```
fix naming to something more appropriate, e.g. `WidgetClickListener`

sxzhang25 replied:
Went with lambda function (suggestion from last time).

rjuang commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/MotionPhotoWidget.java:
```diff
@@ -0,0 +1,340 @@
+                            reader = MotionPhotoReader.open(filename, savedSurface);
+                            Log.d(TAG, "New motion photo reader created");
+                        } catch (IOException | XMPException e) {
+                            e.printStackTrace();
```
replace this line with a more meaningful log. E.g.
```
Log.e(TAG, "Exception occurred while opening file", e);
```

rjuang commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/MotionPhotoWidget.java:
```diff
@@ -0,0 +1,340 @@
+                (double) videoWidth / videoHeight : (double) videoHeight / videoWidth;
+        Log.d(TAG, "asepct ratio: " + aspectRatio);
+        int newVideoWidth, newVideoHeight;
+        if (viewHeight > (int) (viewWidth * aspectRatio)) {
+            // limited by narrow width
+            if (fill) {
+                newVideoWidth = (int) (viewHeight / aspectRatio);
+                newVideoHeight = viewHeight;
+            } else {
+                newVideoWidth = viewWidth;
+                newVideoHeight = (int) (viewWidth * aspectRatio);
+            }
+        } else {
+            // limited by short height
+            if (fill) {
+                newVideoWidth = viewWidth;
+                newVideoHeight = (int) (viewWidth * aspectRatio);
+            } else{
+                newVideoWidth = (int) (viewHeight / aspectRatio);
+                newVideoHeight = viewHeight;
+            }
+        }
```
This will be easier to read if replaced with a private static helper method.

sxzhang25 replied:
(Deleted this method + usages, will use OpenGL to adjust later).

rjuang commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/MotionPhotoWidget.java:
```diff
@@ -0,0 +1,340 @@
+
+        return myState;
+    }
+
```
nit: remove extra blank line

sxzhang25 committed:
Fix formatting.

rjuang commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/BufferProcessor.java:
```diff
@@ -136,9 +142,33 @@ public void process(Bundle messageData) {
+                long renderTimestampNs;
+                long currentTimestampNs = System.nanoTime();
+                if (prevRenderTimestampNs == 0) {
+                    renderTimestampNs = currentTimestampNs + frameDeltaUs * US_TO_NS;
```
Can we define:

```
long frameDeltaNs = frameDeltaUs * US_TO_NS;
```

Then use frameDeltaNs instead of conversions in a bunch of places ?

rjuang commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/MotionPhotoWidget.java:
```diff
@@ -0,0 +1,297 @@
+            @Override
+            public void surfaceChanged(SurfaceHolder holder, int format, int width, int height) {
+                Log.d(TAG, "Surface changed");
+                Log.d(TAG, "Surface changed");
+
```
extra log?

sxzhang25 committed:
Refactor for readability.

- Use nanoseconds for frame delta in BufferProcessor.java
- Remove unneeded log messages and add descriptive error logs
- Rewrite MainActivity.java to play motion photos from raw resource files rather than from sdcard
- Clean up import statements

rjuang commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/BufferProcessor.java:
```diff
@@ -20,6 +20,10 @@
     private static final String TAG = "BufferProcessor";
     private static final long TIMEOUT_US = 1000L;
     private static final long US_TO_NS = 1000L;
+    private static final long THIRTY_FPS_NS = 1_000_000_000 / 30;
```
suffix the 1_000_000_000 with L so that it's interpretted as a long. i.e.,:

```
1_000_000_000L / 30;
```

rjuang replied:
Can we rename this to something like `FALLBACK_FRAME_INTERVAL_NS` and also add a comment above the line to explain what the frame interval is for ? e.g., This is frame interval used if the render timestamp is less than the timestamp of the previous frame rendered.

rjuang commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/MotionPhotoWidget.java:
```diff
@@ -0,0 +1,289 @@
+import java.util.concurrent.Executors;
+
+/**
+ * An Android app widget to set up a video player for a motion photo file.
```
Better comment:

Widget that can load and play motion photos video files.

rjuang commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/MotionPhotoWidget.java:
```diff
@@ -0,0 +1,289 @@
+    }
+
+    public void play() {
+        playProcess = new PlayProcess();
```
should we check that playProcess != null  and cancel it if it still happens to exist? e.g.,

```
if (playProcess != null) {
  playProcess.cancel();
}
playProcess = new PlayProcess();
```

rjuang commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/MotionPhotoWidget.java:
```diff
@@ -0,0 +1,289 @@
+     * @param filename is a string pointing to the motion photo file to play.
+     */
+    @RequiresApi(api = Build.VERSION_CODES.M)
+    public void setFile(String filename) {
+        this.file = new File(filename);
+    }
```
does setFile do anything if the widget is already loaded and/or playing?

sxzhang25 committed:
Fix code formatting and update setFile() behavior in MotionPhotoWidget

- Edit javadoc and variable names in MotionPhotoWidget and BufferProcessor
- Update setFile() so that if a file has already been set for this widget, close the old reader and open up a new reader for the file
- Update main activity to change file on long clicks
- Add Espresso tests to check for correct setFile() behavior
- Save current file URI in widget instance state to preserve video on screen rotation

sxzhang25 committed:
Refactor motion photo widget tests so they comprise an independent test suite (create a new activity, activity layout xml, and test file). Merged conflicts in previously applied patch to accomplish this.
